src/gui/torrentcontentmodel.cpp

```diff
@@ -497,11 +497,45 @@ void TorrentContentModel::setupModelData(const BitTorrent::AbstractFileStorage &
     qDebug("Torrent contains %d files", filesCount);
     m_filesIndex.reserve(filesCount);
 
-    TorrentContentModelFolder *currentParent;
+    class FolderAdaptor
+    {
+    public:
+        using FolderAdaptorPtr = std::shared_ptr<FolderAdaptor>;
```

Comment:
If I read correctly, it seems you aren't really using reference counting. If that is the case, then just store raw pointers.

Reply:
>it seems you aren't really using reference counting.

The adaptors should be auto removed after tree is created.

Reply:
Let me ask another way, would the code be cumbersome if raw pointers were used instead?

Reply:
> Let me ask another way, would the code be cumbersome if raw pointers were used instead?

Then we will have to take care of manually deleting all adapters when they become unnecessary (after creating the tree). Aren't smart pointers created specifically to avoid this?

Reply:
>Then we will have to take care of manually deleting all adapters when they become unnecessary (after creating the tree). Aren't smart pointers created specifically to avoid this?

I can't really comment on how this specific code is better with smart pointers or not...
While I can understand what the old code do easily. I can't really comprehend how the new code manages ptrs, the lifetime of the shared ptr isn't clear to me.
Therefore I'm trying to ask whether it can be simplified, either by using raw ptrs (or even std::unique_ptr) and rely less on shared ptr.

Reply:
>I fail to comprehend how the new code manages ptrs, the lifetime of the shared ptr is hard to follow (especially when some method returns shared ptr).

No such `shared_ptr` goes outside scope of `TorrentContentModel::setupModelData()`.
>Therefore I'm trying to ask whether it can be simplified, either by using raw ptrs (or even `std::unique_ptr`) and rely less on shared ptr.

`std::unique_ptr` will most likely fit but won't it had too many worries with moving back and forth? The raw pointers may even turn out to look better (maybe I just don't like `std::unique_ptr` so much).
> The adaptor class seems so strange to me... [#16183 (comment)](https://github.com/qbittorrent/qBittorrent/pull/16183#discussion_r789697508)

It just provides the similar interface as `TorrentContentModelFolder` to the logic that builds file tree but manages some additional data on top of it that is needed only during tree construction and becomes unused after.
The first option of @jagannatharjun was just kept this data in `TorrentContentModelFolder` for the entire lifetime of the object, which is bad.
The second option (suggested by me) was to have two-level hash table in `TorrentContentModel::setupModelData()` method, but it was found by @jagannatharjun to be insufficiently effective.
Using an adapter solves both problems.

Reply:
> `std::unique_ptr` will most likely fit but won't it had too many worries with moving back and forth?

And it will most likely require to use both `std::unique_ptr` and raw pointers.